related_products_ecommerce/models/sale_order.py

```diff
@@ -0,0 +1,36 @@
+from odoo import api, models
+from odoo.exceptions import ValidationError
+
+
+class SaleOrder(models.Model):
+    _inherit = 'sale.order'
+
+    def _cart_update(self, product_id, line_id=None, add_qty=1, set_qty=0, **kwargs):
+        """
+        - Users can modify the quantity of actual products (manually added products).
+        - Related products will automatically update their quantity based on the actual product's quantity.
+        - Users cannot manually change or remove related products from the cart.
+        """
+        res = super()._cart_update(product_id, line_id, add_qty, set_qty, **kwargs)
+        product = self.env["product.product"].browse(product_id)
+
+        if not product.exists():
+            raise ValidationError(self.env._("Product with ID %s not found") % product_id)
+
+        actual_product_line = self.order_line.filtered(lambda l: l.product_id.id == product_id)
+        actual_qty = actual_product_line.product_uom_qty if actual_product_line else 0
+
+        for related_product in product.product_tmpl_id.related_product_ids:
+            related_variant_id = related_product.product_variant_id.id
+            related_line = self.order_line.filtered(lambda l: l.product_id.id == related_variant_id)
+
+            if actual_qty == 0:
+                if related_line:
+                    related_line.unlink()
+            else:
+                if related_line:
+                    related_line.product_uom_qty = actual_qty
+                else:
+                    self._cart_update(related_variant_id, add_qty=actual_qty, set_qty=actual_qty)
```

Comment:
```suggestion
        related_lines = self.order_line.filtered(lambda l: l.product_id in product.product_tmpl_id.related_product_ids)
        if actual_qty == 0:
            related_lines.unlink()
            return res
        if related_lines:
            related_line.product_uom_qty = actual_qty
        pending_products = product.product_tmpl_id.related_product_ids - related_lines.product_id
        for related_variant_id in pending_products.ids:
            # TODO create a SaleOrderLine
```
Avoid recursion here and create a manual SaleOrderLine